dht/auto_refreshing_sharder.hh

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include "dht/sharder.hh"
+#include "replica/database.hh"
+
+namespace dht {
+
+/// A sharder which uses table's most up to date effective_replication_map in each invocation.
+/// As a result, the time during which a single token metadata version is kept is reduced.
+/// The purpose of this class is having a sharder which can be passed to long-running processes
+/// which do not need a single view on replication for the whole duration of the process,
+/// and do not need to block topology change barriers.
+///
+/// No guarantees are made about how long token metadata version is kept alive.
+/// It may be released before the next call.
+class auto_refreshing_sharder : public dht::sharder {
+    lw_shared_ptr<replica::table> _table;
+    mutable locator::effective_replication_map_ptr _erm;
+    mutable const dht::sharder* _sharder;
+private:
+    void maybe_refresh() const {
+        // FIXME: Make it notification-based.
+        auto erm = _table->get_effective_replication_map();
+        if (!_erm || _erm->get_token_metadata().get_version() != erm->get_token_metadata().get_version()) {
+            _erm = std::move(erm);
+            _sharder = &_erm->get_sharder(*_table->schema());
+        }
+    }
+public:
+    auto_refreshing_sharder(lw_shared_ptr<replica::table> table)
+            : _table(std::move(table))
+    { }
+
+    virtual ~auto_refreshing_sharder() = default;
+
+    virtual unsigned shard_of(const dht::token& token) const override {
+        maybe_refresh();
+        return _sharder->shard_of(token);
```

Comment:
The token might not be owned by us any more, if the stream was abandoned.

Reply:
And this can cause us to reference a null compaction_group. We should check that the new topology is compatible for this token.

We also need a higher-level fence for the entire stream, since holding the token_metadata_ptr doesn't protect us any more.

Reply:
Higher layers are supposed to ensure that onwnership is not lost: https://github.com/scylladb/scylladb/pull/14995#discussion_r1318502410